examples/tutorial.dx

```diff
@@ -0,0 +1,483 @@
+  second dimension has size 10.
+
+:t x
+
+' Once we have an array, we can use it in new comprehensions. For example,
+  let's try to add `5` to each array element. In Python, one might write this as:
+
+' `y = [[x[i][j] for j in range(10)] for i in range(5)]`
+
+' Dex can do something similar. The main superficial difference is the
+  array indexing syntax, which uses `array.i` instead of square brackets for
+  subscripting.
+
+y = for i:(Fin 5). for j:(Fin 10). x.i.j + 5
+
+:p y
+
+' However, we can make this expression nicer. Because `x` has a known array type
+  and `i` and `j` index into that type, Dex can infer the range of the loop.
+  That means that we can safely remove the explicit `Fin` type annotations and
+  get the same result.
+
+y' = for i. for j. x.i.j + 5
+
+' We can further reduce this array by applying array reduction functions like
+  `sum`:
+
+z = for i. sum x.i
+
+' This style of using `for` to construct type-inferred arrays is central to what
+  makes Dex powerful. Let's consider another example. This one produces a list of
+  length 50 in Python.
+
+' `x = [1 for j in range(10) for i in range(5)]`
+
+' The analogous array construct in Dex is written in the following form. It
+  produces a one-dimensional array of 50 elements.
+
+x2 = for (i, j): (Fin 5 & Fin 10). 1
+
+' As before, we can implement "adding 5" to this array using a `for` constructor,
+  enumerating over each of its elements:
+
+y2 = for i. x2.i + 5
+
+' And we can apply array functions to the array:
+
+:p sum x2
+
+' But things start to get interesting when we consider the type of the array.
+  Unlike the Python example, which produces a list of length 50, the Dex array
+  Jmaintains the index type of its construction. In particular, the type of the
+  array remembers the original ranges.
+
+:t x2
+
+'## Typed indexing
+
+' The use of typed indices lets you do really neat things, but it also breaks
+  other things in counterintuitive ways. Dex uses the `.` syntax for array
+  indexing. Critically though, one cannot simply index an array with an integer
+  literal.
+
+r = x.3
+
+' Instead, it is necessary to cast the integer into the index type of the
+  current shape. This type annotation is done with the `@` operator. (If it
+  helps, you can think of array indexing as function application: `a.i` applies
+  array `a` with index `i` just like how `f x` applies function `f` with
+  argument `x`.)
+
+:t x
+
+row = x.(3 @ Fin 5)
+
+:t row
+
+:t row.(5 @ Fin 10)
+
+' This explicit annotation is a bit verbose, but it is rarely necessary in
+  practice. Most of the time, the `for` construct can infer index types.
+  That's why we didn't have to use any `@` symbols when constructing `y2` above.
+
+' Similarly, you cannot use indices as integers as you might be used to. It is
+  necessary to explicitly annotate index types.
+
+x4 = for i:(Fin 5). for j:(Fin 10). i + j
+
+x4 = for i:(Fin 5). for j:(Fin 10). (ordinal i) + (ordinal j)
```

Comment:
It might be a good idea to support this with a little rationale. The type `Fin 10` means that it's an integer between 0 and 10, but this set is not closed under addition, which is why you have to cast to integers first!